### PR TITLE
kg: add new handler for rendering the topology graph via the metrics webserver

### LINKED ISSUE #23: feat issue: Web view of topology graph
Set a controller with http for web view of topology graph.

### PR DESCRIPTION
**Pull request by stv0g**
Closes #23

@squat , @leonnicolas: I am not too happy about the code-duplication for creating a `mesh.Topology` instance.

At the same time, I am lacking the deeper insight into Kilos code-base to understand where we could put this code.
I was thinking about a function `mesh.Backend.Topology()` or `mesh.Mesh.Topology()`.
I guess the main question would be then weather we want to query the current topology directly from the backend or if we take the current state from the mesh object?

### CODE REVIEW

File: cmd/kg/handlers.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		return
+	}
+
+	var hostname string
+	subnet := mesh.DefaultKiloSubnet
+	nodes := make(map[string]*mesh.Node)
+	for _, n := range ns {
+		if n.Ready() {
+			nodes[n.Name] = n
+			hostname = n.Name
```
Comment (squat) [inline cmd/kg/handlers.go:52]: We always know the name of the node running this process, since it's provided as a command line argument to `kg`. I would saw we should use this value always. It will also help future-proof the code I think

File: cmd/kg/handlers.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		if n.WireGuardIP != nil {
+			subnet = n.WireGuardIP
+		}
+	}
+	subnet.IP = subnet.IP.Mask(subnet.Mask)
+	if len(nodes) == 0 {
+		http.Error(w, "did not find any valid Kilo nodes in the cluster", 500)
```
Comment (squat) [inline cmd/kg/handlers.go:60]:
```suggestion
		http.Error(w, "did not find any valid Kilo nodes in the cluster", http.StatusInternalServerError)
```

File: cmd/kg/handlers.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	peers := make(map[string]*mesh.Peer)
+	for _, p := range ps {
+		if p.Ready() {
+			peers[p.Name] = p
+		}
+	}
+	topo, err := mesh.NewTopology(nodes, peers, h.granularity, hostname, 0, []byte{}, subnet, nodes[hostname].PersistentKeepalive, nil)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to create topology: %v", err), 500)
```
Comment (squat) [inline cmd/kg/handlers.go:71]: same here
```suggestion
		http.Error(w, fmt.Sprintf("failed to create topology: %v", err), http.StatusInternalServerError)
```

File: cmd/kg/handlers.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+		return
+	}
+
+	dot, err := topo.Dot()
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to generate graph: %v", err), 500)
```
Comment (squat) [inline cmd/kg/handlers.go:77]: once more :)

```suggestion
		http.Error(w, fmt.Sprintf("failed to generate graph: %v", err), http.StatusInternalServerError)
```

File: cmd/kg/handlers.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+	err = stdin.Close()
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	output, err := command.Output()
+	if err != nil {
+		http.Error(w, fmt.Sprintf("unable to execute dot: %v (is graphviz package installed?)", err), 500)
```
Comment (squat) [inline cmd/kg/handlers.go:115]: I think this might be too much context for the user. Maybe we should rather log the question about graphviz and present a simpler error message to the HTTP client

File: cmd/kg/handlers.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	command.Stderr = os.Stderr
+
+	stdin, err := command.StdinPipe()
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	_, err = io.Copy(stdin, buf)
+	if err != nil {
```
Comment (squat) [inline cmd/kg/handlers.go:102]: By convention in this code base we try to keep error checks on a single line, e.g.
```suggestion
	if _, err := io.Copy(stdin, buf); err != nil {
```

File: cmd/kg/handlers.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	err = stdin.Close()
+	if err != nil {
```
Comment (squat) [inline cmd/kg/handlers.go:108]: same here
```suggestion
	if err := stdin.Close(); err != nil {
```

File: cmd/kg/handlers.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	"io"
+	"mime"
+	"net/http"
+	"os"
+	"os/exec"
+
+	"github.com/squat/kilo/pkg/mesh"
+)
+
+type GraphHandler struct {
```
Comment (squat) [inline cmd/kg/handlers.go:29]: This does not need to be an exported type, right? let's rather keep this as 
```suggestion
type graphHandler struct {
```

File: cmd/kg/handlers.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+	mimeType := mime.TypeByExtension("." + format)
+	if mimeType == "" {
+		mimeType = "application/octet-stream"
+	}
+
+	w.Write(output)
+}
+
+type HealthHandler struct {
```
Comment (squat) [inline cmd/kg/handlers.go:127]: same here.
```suggestion
type healthHandler struct {
```

Comment (squat) [reply cmd/kg/handlers.go:127]: Thinking about this again, since this handler does not actually do any health checking (today) we can probably simplify this and remove the need for a struct. Instead, we should declare `healthHandler` to be a func that matches https://pkg.go.dev/net/http#HandlerFunc and register that with the `mux` using `mux.HandleFunc()`

File: cmd/kg/handlers.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	if format == "" {
+		format = "png"
+	} else if format == ".dot" || format == ".gv" {
+		// If the raw dot data is requested, return it as string.
+		// This allows client-side rendering rather than server-side.
+		w.Write(buf.Bytes())
+		return
+	}
+
+	command := exec.Command("dot", "-T"+format)
```
Comment (squat) [inline cmd/kg/handlers.go:92]: I've found that the `circo` generator produces much more helpful graphs for Kilo than `dot` tends to. Can we default to using this?

Comment (squat) [reply cmd/kg/handlers.go:92]: Also, appending `format` to the command line here looks like a security issue since users could pass unvalidated arguments/commands to `exec`. I think we should have a `switch` statement with cases for the supported formats and only accept those.

File: cmd/kg/handlers.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	dot, err := topo.Dot()
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to generate graph: %v", err), 500)
+	}
+
+	buf := bytes.NewBufferString(dot)
+
+	format := r.URL.Query().Get("format")
+	if format == "" {
+		format = "png"
```
Comment (squat) [inline cmd/kg/handlers.go:84]: I think SVG might be a nicer default format; personally this is the only format I use when viewing graphs from Kilo.

**Review by squat — COMMENTED** [review]
This is a nice feature! Thanks for working on this, @stv0g!!!
Do you think you could also add an e2e test that at a minimum validates that the response contains some SVG XML? This would be sufficient IMO. Let me know if you need help here, or we can try to tackle it in a follow-up.

File: cmd/kg/handlers.go (hunk 1 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done :) I now pass the hostname and the Kilo subnet to the handler.

However, wouldnt it be nicer if we export those fields from the Mesh and only pass the mesh to the handler?

File: cmd/kg/handlers.go (hunk 2 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 3 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 4 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 6 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 7 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 8 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 9 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done. I only did this to keep it consistent with the graph handler and for the future in which the health handler should probably do some more thorough checks

File: cmd/kg/handlers.go (hunk 11 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 1 above)
Comment (squat) [reply cmd/kg/handlers.go]: yes, something in this direction could be nice. Maybe we want the mesh to actually provide a method for getting ready nodes and peers? We need this logic in a bunch of places, like kgctl, so it could be nice to reuse

File: cmd/kg/handlers.go (hunk 10 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: cmd/kg/handlers.go (hunk 5 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

**Comment by stv0g** [comment]
Hi @squat,

thanks for the review :) I think i've addressed most of them.
I just dont like how I need to pass hostname and subnet in addition to the mesh to the graph handler since they are already members of the mesh type itself.

But it seems like Mesh does not really export anything. So I am not too sure how you want to handle this.

I also gave the e2e tests a try in my last commit.

Cheers,
Steffen

File: e2e/handlers.sh [hunk 12]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Comment (squat) [inline e2e/handlers.sh:1]: great work here!

File: e2e/handlers.sh [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+setup_suite() {
+	# shellcheck disable=SC2016
+	_kubectl patch ds -n kube-system kilo -p '{"spec": {"template":{"spec":{"containers":[{"name":"kilo","args":["--hostname=$(NODE_NAME)","--create-interface=false","--kubeconfig=/etc/kubernetes/kubeconfig","--mesh-granularity=full"]}]}}}}'
+	block_until_ready_by_name kube-system kilo-userspace 
+	_kubectl wait pod -l app.kubernetes.io/name=adjacency --for=condition=Ready --timeout 3m
+}
+
+test_graph_handler() {
+    assert "curl_pod http://10.4.0.1:1107/graph?format=svg&layout=circo | grep -q '<svg'"
```
Comment (squat) [inline e2e/handlers.sh:13]: Great work with all of these tests!

File: e2e/handlers.sh [hunk 14]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+# shellcheck disable=SC1091
+. lib.sh
+
+setup_suite() {
+	# shellcheck disable=SC2016
+	_kubectl patch ds -n kube-system kilo -p '{"spec": {"template":{"spec":{"containers":[{"name":"kilo","args":["--hostname=$(NODE_NAME)","--create-interface=false","--kubeconfig=/etc/kubernetes/kubeconfig","--mesh-granularity=full"]}]}}}}'
```
Comment (squat) [inline e2e/handlers.sh:7]: We can remove this patch and re-use the default cluster configuration. Unless you really want to ensure that we necessarily use full-mesh.

File: e2e/handlers.sh (hunk 13 above)
Comment (squat) [reply e2e/handlers.sh:13]: If we add a "should ..." comment to the end of the assert statement then we can get a bit more insight about failures when tests don't pass

File: e2e/handlers.sh [hunk 15]
```diff
@@ -4,14 +4,13 @@
...
 	block_until_ready_by_name kube-system kilo-userspace 
 	_kubectl wait pod -l app.kubernetes.io/name=adjacency --for=condition=Ready --timeout 3m
 }
 
 test_graph_handler() {
     assert "curl_pod http://10.4.0.1:1107/graph?format=svg&layout=circo | grep -q '<svg'"
-    assert "curl_pod http://10.4.0.1:1107/graph&layout=circo | grep -q '<svg'"
+    assert "curl_pod http://10.4.0.1:1107/graph?layout=circo | grep -q '<svg'"
     assert "curl_pod http://10.4.0.1:1107/graph | grep -q '<svg'"
     assert _not "curl_pod http://10.4.0.1:1107/graph?format=svg&layout=fake | grep -q '<svg'"
```
Comment (leonnicolas) [inline e2e/handlers.sh:15]: can we use assert_fail here?

File: e2e/handlers.sh (hunk 14 above)
Comment (stv0g) [reply e2e/handlers.sh]: Done

File: e2e/handlers.sh (hunk 13 above)
Comment (stv0g) [reply e2e/handlers.sh]: Done

File: e2e/handlers.sh (hunk 15 above)
Comment (stv0g) [reply e2e/handlers.sh]: Done

File: cmd/kg/handlers.go [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+		// If the raw dot data is requested, return it as string.
+		// This allows client-side rendering rather than server-side.
+		w.Write(buf.Bytes())
+		return
+	} else if format == "" {
+		format = "svg"
+	} else if format != "png" && format != "bmp" && format != "fig" && format != "gif" && format != "json" && format != "ps" {
+		http.Error(w, "unsupported format", http.StatusInternalServerError)
+		return
+	}
```
Comment (leonnicolas) [inline cmd/kg/handlers.go:89]: The test fails now because `svg` is an unsupported format.
Maybe a switch case makes this more readable and default to unsupported?

File: cmd/kg/handlers.go [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+	layout := r.URL.Query().Get("layout")
+	if layout == "" {
+		layout = "circo"
+	} else if layout != "dot" && layout != "neato" && format != "twopi" && format != "fdp" {
+		http.Error(w, "unsupported layout", http.StatusInternalServerError)
+		return
+	}
```
Comment (leonnicolas) [inline cmd/kg/handlers.go:97]: same thing here. This makes circo an unsupported layout.

File: cmd/kg/handlers.go [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+	if err != nil {
+		http.Error(w, "unable to render graph", http.StatusInternalServerError)
+		return
+	}
+
+	mimeType := mime.TypeByExtension("." + format)
+	if mimeType == "" {
+		mimeType = "application/octet-stream"
+	}
+
```
Comment (leonnicolas) [inline cmd/kg/handlers.go:128]: Don't we need something like `w.Header().Add("content-type", mimeType)` or how is the mimeType variable working?

File: cmd/kg/handlers.go (hunk 16 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Oh, I was still in my Python-world and was assuming Go has no switch statement :D

Done :)

File: cmd/kg/handlers.go (hunk 18 above)
Comment (stv0g) [reply cmd/kg/handlers.go:138]: Done

File: cmd/kg/handlers.go (hunk 17 above)
Comment (stv0g) [reply cmd/kg/handlers.go]: Done

File: e2e/handlers.sh [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+. lib.sh
+
+setup_suite() {
+	# shellcheck disable=SC2016
+	block_until_ready_by_name kube-system kilo-userspace 
+	_kubectl wait pod -l app.kubernetes.io/name=adjacency --for=condition=Ready --timeout 3m
+}
+
+test_graph_handler() {
+    assert "curl_pod http://10.4.0.1:1107/graph?format=svg&layout=circo | grep -q '<svg'" "graph handler should produce SVG output"
```
Comment (leonnicolas) [inline e2e/handlers.sh:12]:
```suggestion
    assert "curl_pod 'http://10.4.0.1:1107/graph?format=svg&layout=circo' | grep -q '<svg'" "graph handler should produce SVG output"
```
I think the `&` is the trouble here.

File: e2e/handlers.sh [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    assert_fail "curl_pod http://10.4.0.1:1107/graph?format=fake | grep -q '<svg'" "graph handler should reject invalid format"
+}
+
+test_health_handler() {
+    assert "curl_pod http://10.4.0.1:1107/health" "health handler should return a status code of 200"
+}
+
+test_metrics_handler() {
+    assert "curl_pod http://10.4.0.1:1107/metrics" "metrics handler should return a status code of 200"
+    assert "(( $(curl_pod http://10.4.0.1:1107/metrics | egrep ^kilo_nodes | cut -d " " -f 2) > 0 ))" "metrics handler should provide metric: kilo_nodes > 0"
```
Comment (leonnicolas) [inline e2e/handlers.sh:25]:
```suggestion
    assert "(( $(curl_pod http://10.4.0.1:1107/metrics | grep -E ^kilo_nodes | cut -d " " -f 2) > 0 ))" "metrics handler should provide metric: kilo_nodes > 0"
```
to avoid the shellcheck warning

**Comment by leonnicolas** [comment]
When I generate a png with `curl '<Kilo Pod IPr>:1107/graph&format=png' -o g.png'`, the circles look as expected, but all text is not rendered correctly. Looks like a formatting issue or the font is not known.
When I generate the png with `curl  '<Kilo Pod IP>:1107/graph&format=dot' | crico -Tpng` everything look great.
Do you have a better idea why this is not working?

**Comment by stv0g** [comment]
Thanks @leonnicolas for looking into it. I had trouble with my dev environment to properly debug the issues here.

> When I generate a png with curl '<Kilo Pod IPr>:1107/graph&format=png' -o g.png', the circles look as expected, but all text is not rendered correctly. Looks like a formatting issue or the font is not known.

Thats due to missing fonts in your containers. The PR includes a change in the Dockerfile to install some standard font:

https://github.com/squat/kilo/pull/214/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557

**Comment by leonnicolas** [comment]
> Thats due to missing fonts in your containers. The PR includes a change in the Dockerfile to install some standard font:

I was building the container images from your branch :(

**Comment by leonnicolas** [comment]
g

I checked out your branch and run the e2e test. The svg is fine, but the png looks like this.

Either we fix this, or disable the png (and other formats) options.

Installing font-noto fixed the issue.

**Comment by stv0g** [comment]
@leonnicolas thanks for investigating. I added font-noto to the Dockerfile

File: Dockerfile [hunk 21]
```diff
@@ -11,7 +11,7 @@ ARG GOARCH
 ARG ALPINE_VERSION=v3.12
 LABEL maintainer="squat <lserven@gmail.com>"
 RUN echo -e "https://alpine.global.ssl.fastly.net/alpine/$ALPINE_VERSION/main\nhttps://alpine.global.ssl.fastly.net/alpine/$ALPINE_VERSION/community" > /etc/apk/repositories && \
-    apk add --no-cache ipset iptables ip6tables wireguard-tools
+    apk add --no-cache ipset iptables ip6tables wireguard-tools graphviz font-bitstream-type1 font-noto
```
Comment (squat) [inline Dockerfile:14]: Can we do this without font-bitstream-type1 and just noto? It would be nice to keep the container with just the essentials

**Comment by squat** [comment]
I went ahead and added a commit removing the extra font. When testing this with the e2e tests everything works. I'm able to spin up a kind cluster, port forward to a Kilo pod and open `http://localhost:1107/graph?format=png` in the browser to produce the following image.
image
I think this is ready to be merged!

**Review by squat — APPROVED** [review]

**Comment by leonnicolas** [comment]
See 1b5ad035d934368491640c66a4dfa2ddddb9ebb2 for the squashed commit in upstream.
